File: src/simulation/power/simpleSolarPanel/simpleSolarPanel.py
import torch
import torch.nn as nn
from typing import Optional

__all__ = ['SimpleSolarPanel']

# 天文常量 (单位: 米)
AU = 149597870700.0  # 1 AU = 149,597,870,700 米
SOLAR_FLUX_EARTH = 1367.0  # 地球附近的太阳通量 (W/m²)


class SimpleSolarPanel(nn.Module):
    """
    PyTorch 实现的太阳能电池板模型

    功能:
    - 计算太阳相对于航天器的位置
    - 计算太阳在航天器本体坐标系中的方向
    - 计算电池板的投影面积
    - 计算太阳距离因子
    - 计算电池板产生的净功率

    参数:
        nHat_B: 面板法向量 (在航天器本体坐标系中)
        panel_area: 面板面积 (m²)
        panel_efficiency: 面板效率 (0-1)
    """

    def __init__(
            self,
            nHat_B: torch.Tensor,
            panel_area: float,
            panel_efficiency: float
    ) -> None:
        """
        初始化太阳能电池板

        参数:
            nHat_B: 面板法向量 (在航天器本体坐标系中), shape [3]
            panel_area: 面板面积 (m²)
            panel_efficiency: 面板效率 (0-1)
        """
        super().__init__()

        # 验证输入参数
        if panel_area <= 0:
            raise ValueError("panel_area must be a positive value")
        if panel_efficiency <= 0 or panel_efficiency > 1:
            raise ValueError("panel_efficiency must be in (0, 1]")
        if torch.norm(nHat_B) < 1e-5:
            raise ValueError("nHat_B must be a non-zero vector")

        # 注册参数
        self.register_buffer('nHat_B', nHat_B / torch.norm(nHat_B))  # 单位化法向量
        self.panel_area = panel_area
        self.panel_efficiency = panel_efficiency

        # 初始化状态变量
        self.shadow_factor: torch.Tensor = torch.tensor(1.0)
        self.projected_area: torch.Tensor = torch.tensor(0.0)
        self.sun_distance_factor: torch.Tensor = torch.tensor(0.0)

    def forward(
            self,
            r_BN_N: torch.Tensor,
            r_SN_N: torch.Tensor,
            sigma_BN: torch.Tensor,
            shadow_factor: Optional[torch.Tensor] = None
    ) -> torch.Tensor:
        """
        计算太阳能电池板的输出功率

        参数:
            r_BN_N: 航天器位置 (在惯性系中), shape [batch_size, 3]
            r_SN_N: 太阳位置 (在惯性系中), shape [batch_size, 3]
            sigma_BN: 航天器姿态 (MRP参数), shape [batch_size, 3]
            shadow_factor: 日食因子 (可选), shape [batch_size]

        返回:
            net_power: 净输出功率, shape [batch_size]
        """
        # 处理日食因子
        if shadow_factor is None:
            shadow_factor = torch.ones(r_BN_N.shape[0], device=r_BN_N.device)

        # 计算太阳相关数据
        self._compute_sun_data(r_BN_N, r_SN_N, sigma_BN)

        # 计算太阳功率因子
        sun_power_factor = SOLAR_FLUX_EARTH * self.sun_distance_factor * shadow_factor

        # 计算净功率
        net_power = sun_power_factor * self.projected_area * self.panel_efficiency

        return net_power

    def _compute_sun_data(
            self,
            r_BN_N: torch.Tensor,
            r_SN_N: torch.Tensor,
            sigma_BN: torch.Tensor
    ) -> None:
        """
        计算太阳相关数据

        参数:
            r_BN_N: 航天器位置 (在惯性系中), shape [batch_size, 3]
            r_SN_N: 太阳位置 (在惯性系中), shape [batch_size, 3]
            sigma_BN: 航天器姿态 (MRP参数), shape [batch_size, 3]
        """
        # 计算太阳相对于航天器的向量
        r_SB_N = r_SN_N - r_BN_N

        # 计算太阳方向的单位向量 (在惯性系中)
        r_SB_norm = torch.norm(r_SB_N, dim=1, keepdim=True)
        sHat_N = r_SB_N / (r_SB_norm + 1e-10)  # 添加小值防止除零

        # 计算惯性系到本体坐标系的旋转矩阵 (从MRP)
        dcm_BN = self._mrp_to_dcm(sigma_BN)

        # 将太阳方向转换到本体坐标系
        sHat_B = torch.bmm(dcm_BN, sHat_N.unsqueeze(-1)).squeeze(-1)

        # 计算投影面积
        cos_theta = torch.sum(sHat_B * self.nHat_B, dim=1)
        projected_area = self.panel_area * cos_theta
        projected_area = torch.clamp(projected_area, min=0.0)  # 负值设为零

        # 计算太阳距离因子 - 修复了括号问题
        sun_distance_factor = (AU ** 2) / (r_SB_norm.squeeze(1) ** 2)

        # 更新状态
        self.projected_area = projected_area
        self.sun_distance_factor = sun_distance_factor

    def _mrp_to_dcm(self, sigma: torch.Tensor) -> torch.Tensor:
        """
        将修正罗德里格斯参数(MRP)转换为方向余弦矩阵(DCM)

        参数:
            sigma: MRP参数, shape [batch_size, 3]

        返回:
            dcm: 方向余弦矩阵, shape [batch_size, 3, 3]
        """
        batch_size = sigma.shape[0]
        sigma_squared = torch.sum(sigma ** 2, dim=1, keepdim=True)

        # 叉乘矩阵
        sigma_tilde = torch.zeros(batch_size, 3, 3, device=sigma.device)
        sigma_tilde[:, 0, 1] = -sigma[:, 2]
        sigma_tilde[:, 0, 2] = sigma[:, 1]
        sigma_tilde[:, 1, 0] = sigma[:, 2]
        sigma_tilde[:, 1, 2] = -sigma[:, 0]
        sigma_tilde[:, 2, 0] = -sigma[:, 1]
        sigma_tilde[:, 2, 1] = sigma[:, 0]
        """
        [ 0     -σ₃    σ₂ ]
        [ σ₃     0     -σ₁]
        [ -σ₂    σ₁     0 ]
        """
        # 计算DCM
        denominator = (1 + sigma_squared) ** 2
        term1 = 8 * torch.bmm(sigma_tilde, sigma_tilde)
        term2 = 4 * (1 - sigma_squared) * sigma_tilde

        # 批量单位矩阵
        eye = torch.eye(3, device=sigma.device).unsqueeze(0).repeat(batch_size, 1, 1)

        dcm = eye + (term1 - term2) / denominator.view(-1, 1, 1)

        return dcm

    def reset(self) -> None:
        """重置状态变量"""
        self.shadow_factor = torch.tensor(1.0)
        self.projected_area = torch.tensor(0.0)
        self.sun_distance_factor = torch.tensor(0.0)

File: src/simulation/power/simpleSolarPanel/simpleSolarPanel.cpp
//
// simpleSolarPanel.cpp
// 太阳能电池板模型实现
// 核心功能：根据太阳位置、航天器状态和面板特性计算输出功率
// 关键算法：太阳矢量投影 + 距离衰减 + 阴影因子
//

#include <math.h>                   // 基础数学函数
#include "simpleSolarPanel.h"       // 类声明
#include "architecture/utilities/rigidBodyKinematics.h"  // 刚体运动学工具
#include "architecture/utilities/astroConstants.h"       // 天文常量 (AU, SOLAR_FLUX_EARTH)
#include "architecture/utilities/avsEigenSupport.h"      // Eigen向量支持
#include "architecture/utilities/avsEigenMRP.h"          // MRP姿态表示支持

// 构造函数：初始化默认参数
SimpleSolarPanel::SimpleSolarPanel() {
    // 初始化输出功率为0
    this->nodePowerOut = 0.0;
    
    // 面板参数初始化为无效值（强制用户设置）
    this->panelArea = -1;           // 面积必须>0
    this->panelEfficiency = -1;     // 效率必须>0
    
    // 法向量初始化为零向量（需用户设置）
    this->nHat_B.setZero();         // Eigen::Vector3d(0,0,0)
    
    // 阴影因子默认无阴影
    this->shadowFactor = 1;         // 1.0=完全光照, 0.0=完全阴影
    
    return;
}

// 析构函数：空实现（无动态资源需释放）
SimpleSolarPanel::~SimpleSolarPanel() {
    return;
}

/*! 
 * 重置函数：初始化或重置模块状态
 * @param CurrentClock 当前仿真时钟（未使用）
 * 
 * 功能：
 * 1. 验证关键参数有效性
 * 2. 归一化法向量
 * 3. 检查输入消息连接
 */
void SimpleSolarPanel::customReset(uint64_t CurrentClock) {
    // 重置阴影因子为默认值（无阴影）
    this->shadowFactor = 1.0;

    // === 参数校验 ===
    // 检查面板面积是否有效（必须正数）
    if (this->panelArea < 0.0) {
        bskLogger.bskLog(BSK_ERROR, "The panelArea must be a positive value");
    }
    
    // 检查面板效率是否有效（0.0~1.0）
    if (this->panelEfficiency < 0.0) {
        bskLogger.bskLog(BSK_ERROR, "The panelEfficiency variable must be a positive value");
    }
    
    // 检查法向量并归一化
    // 注意：使用0.1阈值避免浮点精度问题（非零检查）
    if (this->nHat_B.norm() > 0.1) {
        // 归一化法向量（确保为单位向量）
        this->nHat_B.normalize();
    } else {
        // 法向量过小或为零，报错
        bskLogger.bskLog(BSK_ERROR, "The nHat_B must be set to a non-zero vector");
    }
    
    // === 输入消息校验 ===
    // 检查太阳位置消息是否连接
    if (!this->sunInMsg.isLinked()) {
        bskLogger.bskLog(BSK_ERROR, "simpleSolarPanel.sunInMsg was not linked.");
    }
    
    // 检查航天器状态消息是否连接
    if (!this->stateInMsg.isLinked()) {
        bskLogger.bskLog(BSK_ERROR, "simpleSolarPanel.stateInMsg was not linked.");
    }

    return;
}

/*!
 * 消息读取函数：从输入消息获取数据
 * @return 总是返回true（未实现错误处理）
 * 
 * 功能：
 * 1. 读取太阳位置消息
 * 2. 读取航天器状态消息
 * 3. 读取日食消息（可选）
 */
bool SimpleSolarPanel::customReadMessages() {
    // 初始化消息数据（安全默认值）
    this->sunData = sunInMsg.zeroMsgPayload;    // 太阳位置 (0,0,0)
    this->stateCurrent = stateInMsg.zeroMsgPayload; // 航天器状态 (零状态)

    // 读取太阳位置消息（如果已连接）
    if (this->sunInMsg.isLinked()) {
        this->sunData = this->sunInMsg();  // 获取最新消息
    }
    
    // 读取航天器状态消息（如果已连接）
    if (this->stateInMsg.isLinked()) {
        this->stateCurrent = this->stateInMsg(); // 获取最新消息
    }
    
    // 读取日食消息（可选，影响阴影因子）
    if (this->sunEclipseInMsg.isLinked()) {
        EclipseMsgPayload sunVisibilityFactor; // 阴影因子消息
        sunVisibilityFactor = this->sunEclipseInMsg(); // 获取阴影因子
        this->shadowFactor = sunVisibilityFactor.shadowFactor; // 更新阴影因子
    }
    
    return true; // 总是返回成功（实际应添加错误检查）
}

/*!
 * 设置面板物理参数
 * @param nHat_B     面板法向量（航天器本体坐标系）
 * @param panelArea  面板面积 (m²)
 * @param panelEfficiency 面板效率 (0.0~1.0)
 * 
 * 注意：不会立即归一化法向量（在customReset中处理）
 */
void SimpleSolarPanel::setPanelParameters(Eigen::Vector3d nHat_B, 
                                         double panelArea, 
                                         double panelEfficiency) {
    this->nHat_B = nHat_B;             // 存储法向量
    this->panelArea = panelArea;        // 存储面积
    this->panelEfficiency = panelEfficiency; // 存储效率
    return;
}

/*!
 * 核心计算函数：计算太阳相关数据
 * 
 * 计算内容：
 * 1. 航天器-太阳矢量
 * 2. 太阳方向单位矢量（本体系）
 * 3. 面板投影面积
 * 4. 日距衰减因子
 */
void SimpleSolarPanel::computeSunData() {
    // 变量声明
    Eigen::Vector3d r_SB_N;         // 太阳到航天器的矢量 (惯性系)
    Eigen::Vector3d sHat_N;         // 太阳方向单位矢量 (惯性系)
    Eigen::Matrix3d dcm_BN;         // 惯性系到本体系的旋转矩阵
    
    // 从消息中提取数据（转换为Eigen格式）
    Eigen::Vector3d r_BN_N = cArray2EigenVector3d(this->stateCurrent.r_BN_N);       // 航天器位置
    Eigen::Vector3d r_SN_N = cArray2EigenVector3d(this->sunData.PositionVector);    // 太阳位置
    Eigen::MRPd sigma_BN = cArray2EigenMRPd(this->stateCurrent.sigma_BN);           // 姿态MRP

    // === 计算太阳方向矢量 ===
    // 矢量计算：r_SB_N = r_SN_N - r_BN_N (太阳->航天器)
    r_SB_N = r_SN_N - r_BN_N;
    
    // 归一化得到单位矢量（防止零除错误）
    if (r_SB_N.norm() > 1e-10) {
        sHat_N = r_SB_N.normalized(); // 单位太阳矢量 (惯性系)
    } else {
        sHat_N.setZero(); // 异常情况处理
    }

    // === 姿态转换 ===
    // 从MRP获取旋转矩阵（本体系到惯性系）
    Eigen::Matrix3d dcm_NB = sigma_BN.toRotationMatrix();
    // 转置得到惯性系到本体系的旋转矩阵
    dcm_BN = dcm_NB.transpose();
    
    // 将太阳矢量转换到本体系
    Eigen::Vector3d sHat_B = dcm_BN * sHat_N;

    // === 计算投影面积 ===
    // 点积公式：投影面积 = 实际面积 × cosθ
    // 其中 θ = 太阳矢量与法向量的夹角
    double cosTheta = sHat_B.dot(this->nHat_B);
    
    // 负值处理（太阳在面板背面时面积为0）
    this->projectedArea = (cosTheta > 0) ? 
                          (this->panelArea * cosTheta) : 0.0;

    // === 计算日距因子 ===
    // 平方反比定律：衰减因子 = (AU/r)^2
    double r_SBNorm = r_SB_N.norm();
    // 防止除零错误（当r很小时）
    if (r_SBNorm > 1e-10) {
        // AU单位转换：天文单位 -> 米
        double au_meters = AU * 1000.0;
        this->sunDistanceFactor = (au_meters * au_meters) / (r_SBNorm * r_SBNorm);
    } else {
        this->sunDistanceFactor = 0.0; // 异常情况
    }
}

/*!
 * 功率评估函数：计算面板输出功率
 * @param powerUsageSimMsg 输出消息（存储计算结果）
 * 
 * 功率公式：
 * Power = SOLAR_FLUX_EARTH × (AU²/r²) × ShadowFactor × (Area × cosθ) × Efficiency
 */
void SimpleSolarPanel::evaluatePowerModel(PowerNodeUsageMsgPayload *powerUsageSimMsg) {
    // 步骤1：计算太阳相关数据
    this->computeSunData(); // 更新projectedArea和sunDistanceFactor
    
    // 步骤2：计算太阳功率因子
    // = 地球轨道太阳通量 × 距离衰减 × 阴影因子
    double sunPowerFactor = SOLAR_FLUX_EARTH * this->sunDistanceFactor * this->shadowFactor;
    
    // 步骤3：计算净输出功率
    // = 太阳功率因子 × 有效面积 × 转换效率
    powerUsageSimMsg->netPower = sunPowerFactor * this->projectedArea * this->panelEfficiency;

    return;
}